pkg/controller/job/job_controller.go

```diff
@@ -1298,36 +1326,51 @@ func (jm *Controller) removeTrackingFinalizerFromPods(ctx context.Context, jobKe
 
 // enactJobFinished adds the Complete or Failed condition and records events.
 // Returns whether the Job was considered finished.
-func (jm *Controller) enactJobFinished(job *batch.Job, finishedCond *batch.JobCondition) bool {
-	if finishedCond == nil {
+func (jm *Controller) enactJobFinished(job *batch.Job, jobCtx *syncJobCtx) bool {
+	if jobCtx == nil || (jobCtx.failureCondition == nil && jobCtx.successCondition == nil) {
 		return false
 	}
 	if uncounted := job.Status.UncountedTerminatedPods; uncounted != nil {
 		if len(uncounted.Succeeded) > 0 || len(uncounted.Failed) > 0 {
 			return false
 		}
 	}
-	job.Status.Conditions, _ = ensureJobConditionStatus(job.Status.Conditions, finishedCond.Type, finishedCond.Status, finishedCond.Reason, finishedCond.Message, jm.clock.Now())
-	if finishedCond.Type == batch.JobComplete {
-		job.Status.CompletionTime = &finishedCond.LastTransitionTime
+	if jobCtx.failureCondition != nil {
+		job.Status.Conditions, _ = ensureJobConditionStatus(job.Status.Conditions, jobCtx.failureCondition.Type, jobCtx.failureCondition.Status,
+			jobCtx.failureCondition.Reason, jobCtx.failureCondition.Message, jm.clock.Now())
+		return true
 	}
-	return true
+	if jobCtx.successCondition != nil {
+		job.Status.Conditions, _ = ensureJobConditionStatus(job.Status.Conditions, jobCtx.successCondition.Type, jobCtx.successCondition.Status,
+			jobCtx.successCondition.Reason, jobCtx.successCondition.Message, jm.clock.Now())
+		if jobCtx.successCondition.Type == batch.JobComplete {
+			job.Status.CompletionTime = &jobCtx.successCondition.LastTransitionTime
+		}
+		return true
+	}
+	return false
```

Comment:
Seeing this I sort of prefer to the version with a single `finishedCondition` field in `jobCtx`. In  case here we have repetition of calling `ensureJobConditionStatus` per condition, which could be prevented with the single `finishedCondition`.

In cases we need to discriminate we could check the condition or use helper functions, such as `isSuccessCondition`.

Reply:
Uhm, it sounds reasonable. Let me revert a commit for the introducing the `successCondition`, and then let's introduce the `isSuccessCondition` helper.

Reply:
Done.